pkg/runner/step_context.go

```diff
@@ -185,87 +185,106 @@ func (sc *StepContext) setupEnv(ctx context.Context) (ExpressionEvaluator, error
+	// jobs can receive context values, so we interpolate
+	step.WorkingDirectory = rc.ExprEval.Interpolate(step.WorkingDirectory)
 
-		if _, err = script.WriteString(run); err != nil {
-			return err
-		}
-		scriptName := fmt.Sprintf("workflow/%s", step.ID)
-
-		// Reference: https://github.com/actions/runner/blob/8109c962f09d9acc473d92c595ff43afceddb347/src/Runner.Worker/Handlers/ScriptHandlerHelpers.cs#L47-L64
-		// Reference: https://github.com/actions/runner/blob/8109c962f09d9acc473d92c595ff43afceddb347/src/Runner.Worker/Handlers/ScriptHandlerHelpers.cs#L19-L27
-		runPrepend := ""
-		runAppend := ""
-		scriptExt := ""
-		switch step.Shell {
-		case "bash", "sh":
-			scriptExt = ".sh"
-		case "pwsh", "powershell":
-			scriptExt = ".ps1"
-			runPrepend = "$ErrorActionPreference = 'stop'"
-			runAppend = "if ((Test-Path -LiteralPath variable:/LASTEXITCODE)) { exit $LASTEXITCODE }"
-		case "cmd":
-			scriptExt = ".cmd"
-			runPrepend = "@echo off"
-		case "python":
-			scriptExt = ".py"
-		}
-
-		scriptName += scriptExt
-		run = runPrepend + "\n" + run + "\n" + runAppend
-
-		log.Debugf("Wrote command '%s' to '%s'", run, scriptName)
-		scriptPath := fmt.Sprintf("%s/%s", rc.Config.ContainerWorkdir(), scriptName)
-
-		if step.Shell == "" {
-			step.Shell = rc.Run.Job().Defaults.Run.Shell
-		}
-		if step.Shell == "" {
-			step.Shell = rc.Run.Workflow.Defaults.Run.Shell
-		}
-		if rc.Run.Job().Container() != nil {
-			if rc.Run.Job().Container().Image != "" && step.Shell == "" {
-				step.Shell = "sh"
-			}
-		}
-		scCmd := step.ShellCommand()
+	// but top level keys in workflow file like `defaults` or `env` can't
+	if step.WorkingDirectory == "" {
+		step.WorkingDirectory = rc.Run.Workflow.Defaults.Run.WorkingDirectory
+	}
+}
 
-		var finalCMD []string
-		if step.Shell == "pwsh" || step.Shell == "powershell" {
-			finalCMD = strings.SplitN(scCmd, " ", 3)
-		} else {
-			finalCMD = strings.Fields(scCmd)
-		}
+func (sc *StepContext) setupShell() {
+	rc := sc.RunContext
+	step := sc.Step
 
-		for k, v := range finalCMD {
-			if strings.Contains(v, `{0}`) {
-				finalCMD[k] = strings.Replace(v, `{0}`, scriptPath, 1)
-			}
+	if step.Shell == "" {
+		step.Shell = rc.Run.Job().Defaults.Run.Shell
+	}
+
+	step.Shell = rc.ExprEval.Interpolate(step.Shell)
+
+	if step.Shell == "" {
+		step.Shell = rc.Run.Workflow.Defaults.Run.Shell
+	}
+
+	// current GitHub Runner behaviour is that default is `sh`,
+	// but if it's not container it validates with `which` command
+	// if `bash` is available, and provides `bash` if it is
+	// for now I'm going to leave below logic, will address it in different PR
+	// https://github.com/actions/runner/blob/9a829995e02d2db64efb939dc2f283002595d4d9/src/Runner.Worker/Handlers/ScriptHandler.cs#L87-L91
+	if rc.Run.Job().Container() != nil {
+		if rc.Run.Job().Container().Image != "" && step.Shell == "" {
+			step.Shell = "sh"
 		}
+	}
+}
+
+// TODO: Currently we just ignore top level keys, BUT we should return proper error on them
```

Comment:
I'm not sure if I do understand your todo-comment here. Is it related to not having a test workflow with a top-level default shell/workdir?

Reply:
```
// TODO: Currently we just ignore top level keys, BUT we should return proper error on them
// BUTx2 I leave this for when we rewrite act to use actionlint for workflow validation
// so we return proper errors before any execution or spawning containers
// it will error anyway with:
// OCI runtime exec failed: exec failed: container_linux.go:380: starting container process caused: exec: "${{": executable file not found in $PATH: unknown
```

Reply:
the todo is not for `setupShellCommand` but for workflow parser to validate top level keys cannot use contexts

Reply:
Then it looks like the integration tests you've added do not test for: https://github.com/nektos/act/pull/930/files/18c648a0e35be856afe78bb2948ab4e854fa2116#diff-d428984a0cfea053a858f85bae301c213d3f37466ac24fb031ad514676e6955eR216 right?

Reply:
Yes